### ISSUE TITLE
Expression of type 'Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor+QueryingEnumerable`1[System.String]' cannot be used for return type 'System.String'

### ISSUE THREAD

**Issue by arex388** [dotnet/efcore#18145]
I've upgraded a complex .NET Core 2.2 solution to .NET Core 3.0 over the weekend, and I am now running into somewhat random, but completely inexplicable issues when debugging the ASP.NET Core projects.

When I got to only a small set of "Add" pages I am greeted by this exception:

> Expression of type 'Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor+QueryingEnumerable`1[System.String]' cannot be used for return type 'System.String'

If I reload the page, it works correctly. I reload again after that I get an exception. Each reload alternates between the page loading correctly or throwing the exception. This only occurs when debugging from within Visual Studio by F5 or CTRL+F5. When publishing to IIS in release mode, the page just fails to load every time.

I am using EFCore Plus for future queries, and I'm not sure if that's the cause of the problem. I would imagine it's not considering I do get a load every other time, and other more complex pages with many more future queries never have the issue. The pages in question in fact have the simplest projections set up with only one future query...

### Steps to reproduce

Frankly, I have no idea how to reproduce this. All I know is that this was not an issue when the projects were targeting .NET Core 2.2 and EF Core 2.2.

Here's the stack trace:

```
ArgumentException: Expression of type 'Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor+QueryingEnumerable`1[System.String]' cannot be used for return type 'System.String'

System.Linq.Expressions.Expression.ValidateLambdaArgs(Type delegateType, ref Expression body, ReadOnlyCollection<ParameterExpression> parameters, string paramName)
System.Linq.Expressions.Expression.Lambda<TDelegate>(Expression body, string name, bool tailCall, IEnumerable<ParameterExpression> parameters)
System.Linq.Expressions.Expression.Lambda<TDelegate>(Expression body, bool tailCall, IEnumerable<ParameterExpression> parameters)
Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor<TResult>(Expression query)
Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery<TResult>(Expression query, bool async)
Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore<TResult>(IDatabase database, Expression query, IModel model, bool async)
Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler+<>c__DisplayClass9_0<TResult>.<Execute>b__0()
Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore<TFunc>(object cacheKey, Func<Func<QueryContext, TFunc>> compiler)
Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery<TResult>(object cacheKey, Func<Func<QueryContext, TResult>> compiler)
Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute<TResult>(Expression query)
Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute<TResult>(Expression expression)
Z.EntityFramework.Plus.QueryFutureValue<TResult>.GetResultDirectly()
Z.EntityFramework.Plus.QueryFutureBatch.ExecuteQueries()
Z.EntityFramework.Plus.QueryFutureValue<TResult>.get_Value()
lambda_method(Closure , Projection , View , ResolutionContext )

AutoMapperMappingException: Error mapping types. Mapping types: Projection -> View REDACTED.Web.Admin.Features.Countries.Add+Projection -> REDACTED.Web.Admin.Features.Countries.Add+View Type Map configuration: Projection -> View REDACTED.Web.Admin.Features.Countries.Add+Projection -> REDACTED.Web.Admin.Features.Countries.Add+View Destination Member: Country

lambda_method(Closure , Projection , View , ResolutionContext )
lambda_method(Closure , object , object , ResolutionContext )
AutoMapper.Mapper.Map<TDestination>(object source)
REDACTED.Web.AsyncProjectionHandlerBase<TRequest, TProjection, TResult>.GetResultAsync(TRequest request) in AsyncProjectionHanderlBase`TRequest`TProjection`TResult.cs
+
			var result = Mapper.Map<TResult>(projection);
REDACTED.Web.AsyncProjectionHandlerBase<TRequest, TProjection, TResult>.Handle(TRequest request, CancellationToken cancellationToken) in AsyncProjectionHanderlBase`TRequest`TProjection`TResult.cs
+
			return await GetResultAsync(request);
MediatR.Pipeline.RequestPostProcessorBehavior<TRequest, TResponse>.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
MediatR.Pipeline.RequestPreProcessorBehavior<TRequest, TResponse>.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
REDACTED.Web.Admin.Features.Countries.CountriesController.Add() in CountriesController.cs
+
			var view = await Mediator.Send(new Add.ViewQuery());
Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor+TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
System.Threading.Tasks.ValueTask<TResult>.get_Result()
System.Runtime.CompilerServices.ValueTaskAwaiter<TResult>.GetResult()
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask<IActionResult> actionResultValueTask)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
REDACTED.Web.ApplicationUserMiddleware.InvokeAsync(HttpContext context, IApplicationUserServices services) in ApplicationUserMiddleware.cs
+
			await Next(context);
Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Arex388.AspNetCore.HtmlMinifierMiddleware.InvokeAsync(HttpContext context)
Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.Invoke(HttpContext context)
StackExchange.Profiling.MiniProfilerMiddleware.Invoke(HttpContext context) in MiniProfilerMiddleware.cs
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

### Further technical details

EF Core version: 3.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer 3.0
Target framework: .NET Core 3.0
Operating system: Windows 10, 1903
IDE: Visual Studio 2019 16.3.1

**Comment by smitpatel** [comment]
This issue is lacking enough information for us to effectively reproduce. Please post a runnable project/solution or complete code listing that demonstrates the behavior you are seeing.

**Comment by ajcvickers** [comment]
**EF Team Triage:** Closing this issue as the requested additional details have not been provided and we have been unable to reproduce it.

*BTW this is a canned response and may have info or details that do not directly apply to this particular issue. While we'd like to spend the time to uniquely address every incoming issue, we get a lot traffic on the EF projects and that is not practical. To ensure we maximize the time we have to work on fixing bugs, implementing new features, etc. we use canned responses for common triage decisions.*

**Comment by arex388** [comment]
I'm going to comment just for closure.

I was trying to create a small solution to reproduce the issue, but found out I couldn't. I also found out that my IntelliSense came back when working on a new solution, so I decided to re-create the entire solution into a new from scratch .NET Core 3.0 solution. That took three days to copy everything over and I don't get to bill for it.

Anyway, that didn't work, and the issue continued to appear. Over the course of trying to fix it I noticed that when I was using a DbContext Pool, it would be intermittent and throw on every other page refresh (and *only* for certain pages without a discernible reason). If I *only* used DbContext, it would throw *every* page refresh.

As an optimization, which I'll keep, I decided to stop querying for templates on every page load, and instead inject a `TemplateCache` class that pulls all templates from the DB during app initialization. That helped... It made it throw on two page reloads, after two successful page reloads.

Since I was using EF Plus' Future Queries I decided to remove it from the offending pages, and everything works perfectly again. So... It seems to be an incompatibility between EF Core 3.0 (because it worked perfectly fine in 2.2) and EF Plus 3.0.X.

Sadly, that means I loose out on the batched queries that EF Plus was enabling so it will have a rather negative impact on my project. At that point, to be honest, all of the performance improvements you've been adding become irrelevant to me if I have to perform multiple queries instead of one.

It would be nice if EF Core had native query batching for reading data. Batching on write is only half of a solution. Unless you've already snuck in the ability to do that, but as far as I know it doesn't exist.

Now, I'll be off to go and make my project regress a few years but at least perform. // end of rant
